src/io/EventReader.cc

```diff
@@ -0,0 +1,81 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file EventReader.cc
+//---------------------------------------------------------------------------//
+#include "EventReader.hh"
+#include "HepMC3/GenEvent.h"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a filename.
+ */
+EventReader::EventReader(const char* filename, constSPParticleParams params)
+    : params_(std::move(params))
+{
+    // Determine the input file format and construct the appropriate reader
+    input_file_ = HepMC3::deduce_reader(filename);
+    ENSURE(input_file_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Read the primary particles from the event record.
+ */
+EventReader::result_type EventReader::operator()()
+{
+    result_type result;
+
+    while (!input_file_->failed())
+    {
+        // Parse the next event from the record
+        HepMC3::GenEvent gen_event;
+        input_file_->read_event(gen_event);
+
+        // There are no more events
+        if (input_file_->failed())
+        {
+            break;
+        }
```

Comment:
We probably will start using Celeritas for simulating only EM showers in the CMS tracker / ecal. In that sense, fetching only `electrons` and `gammas` from a PHYTIA input seems a reasonable scenario where we could want to use an `operator()` to load a list of selected particles based on their PDG number, instead of parsing the full event.

Reply:
I can't think of a situation myself where we'd want to read the primaries event-by-event, but maybe someone who knows better can. @stognini that's a good point that I hadn't considered -- let me update the reader so we can optionally only load primaries of a certain type.

Reply:
In that case it might be easier to pre-filter the event file. As implemented, `params->find(PDGNumber)` will cause an assertion error if the particle is undefined, so we'd have to consider how to handle that case and to make sure we're not unintentionally leaving out missing particle types.

Reply:
> so we'd have to consider how to handle that case and to make sure we're not unintentionally leaving out missing particle types.

Since the EventReader will only be executed once on the host side, when loading the hepmc/hepevt data, can we just log the skipped particles (including undefined ones) during the pre-filter stage? This way we know what happens, and we should never get the assertion error from `params->find(PDGNumber)`.

Reply:
Good point. For now though, let's keep the event reader structured as-is.